packages/flutter/lib/src/material/navigation_rail.dart

```diff
@@ -0,0 +1,877 @@
+  ///
+  /// When a [NavigationRailDestination] is not selected,
+  /// [unselectedLabelTextStyle] will be used.
+  ///
+  /// The default value is based on the [Theme]'s
+  /// [ThemeData.textTheme.bodyText]. The default color is based on the
+  /// [Theme]'s [ColorScheme.primary].
+  ///
+  /// Properties from this text style,
+  /// or [NavigationRailThemeData.selectedLabelTextStyle] if this is null, are
+  /// merged into the defaults.
+  final TextStyle selectedLabelTextStyle;
+
+  /// The visual properties of the icon in the unselected destination.
+  ///
+  /// If this field is not provided, or provided with any null properties, then
+  /// a copy of the [IconThemeData.fallback] with a custom [NavigationRail]
+  /// specific color will be used.
+  ///
+  /// The default value is Is the [Theme]'s [ThemeData.iconTheme] with a color
+  /// of the [Theme]'s [ColorScheme.onSurface] with an opacity of 0.64.
+  /// Properties from this icon theme, or
+  /// [NavigationRailThemeData.unselectedIconTheme] if this is null, are
+  /// merged into the defaults.
+  final IconThemeData unselectedIconTheme;
+
+  /// The visual properties of the icon in the selected destination.
+  ///
+  /// When a [NavigationRailDestination] is not selected,
+  /// [unselectedIconTheme] will be used.
+  ///
+  /// The default value is Is the [Theme]'s [ThemeData.iconTheme] with a color
+  /// of the [Theme]'s [ColorScheme.primary]. Properties from this icon theme,
+  /// or [NavigationRailThemeData.selectedIconTheme] if this is null, are
+  /// merged into the defaults.
+  final IconThemeData selectedIconTheme;
+
+  /// The smallest possible width for the rail regardless of the destination's
+  /// icon or label size.
+  ///
+  /// The default is 72.
+  ///
+  /// This value also defines the min width and min height of the destinations.
+  ///
+  /// To make a compact rail, set this to 56 and use
+  /// [NavigationRailLabelType.none].
+  final double minWidth;
+
+  /// The final width when the animation is complete for setting [extended] to
+  /// true.
+  ///
+  /// This is only used when [extended] is set to true.
+  ///
+  /// The default value is 256.
+  final double minExtendedWidth;
+
+  /// Returns the animation that controls the [NavigationRail.extended] state.
+  ///
+  /// This can be used to synchronize animations in the [leading] or [trailing]
+  /// widget, such as an animated menu or a [FloatingActionButton] animation.
+  ///
+  /// {@tool snippet}
+  ///
+  /// This example shows how to use this animation to create a
+  /// [FloatingActionButton] that animates itself between the normal and
+  /// extended states of the [NavigationRail].
+  ///
+  /// An instance of `ExtendableFab` would be created for
+  /// [NavigationRail.leading].
+  ///
+  /// ```dart
+  /// import 'dart:ui';
+  ///
+  /// class NavigationRailExtendableFab extends StatelessWidget {
+  ///   @override
+  ///   Widget build(BuildContext context) {
```

Comment:
I think this could be made slightly easier to read and a little more consistent about type declarations:

```dart
///   @override
///   Widget build(BuildContext context) {
///     final Animation<double> animation = NavigationRail.extendedAnimation(context);
///     return AnimatedBuilder(
///       animation: animation,
///       builder: (BuildContext context, Widget child) {
///         return Container(
///           height: 56,
///           padding: EdgeInsets.symmetric(
///             vertical: lerpDouble(0, 6, animation.value),
///           ),
///           child: animation.value == 0
///             ? FloatingActionButton(
///                 child: Icon(Icons.add),
///                 onPressed: () {},
///               )
///             : Align(
///                 alignment: AlignmentDirectional.centerStart,
///                 widthFactor: animation.value,
///                 child: Padding(
///                   padding: const EdgeInsetsDirectional.only(start: 8),
///                   child: FloatingActionButton.extended(
///                     icon: Icon(Icons.add),
///                     label: Text('CREATE'),
///                     onPressed: () {},
///                   ),
///                 ),
///               ),
///         );
///       },
///     );
```
Also: The Container's height probably merits a comment

Reply:
Done